app/services/katello/registration_manager.rb

```diff
@@ -0,0 +1,156 @@
+module Katello
+  class RegistrationManager
+    class << self
+      private :new
+
+      def unregister_host(host, options = {})
+        organization_destroy = options.fetch(:organization_destroy, false)
+        unregistering = options.fetch(:unregistering, false)
+
+        if !organization_destroy && host.subscription_facet.try(:uuid)
+          ::Katello::Resources::Candlepin::Consumer.destroy(host.subscription_facet.uuid)
+        end
+
+        begin
+          ::Katello.pulp_server.extensions.consumer.delete(host.content_facet.uuid) if host.content_facet.try(:uuid)
+        rescue RestClient::ResourceNotFound
+          Rails.logger.error(_("Pulp Consumer %s has already been removed") % host.content_facet.uuid)
+        end
+
+        host.subscription_facet.try(:destroy!)
+
+        if unregistering
+          remove_host_artifacts(host)
+        elsif organization_destroy
+          host.content_facet.try(:destroy!)
+          destroy_host_artifacts(host)
+        else
+          host.content_facet.try(:destroy!)
+          unless host.destroy
+            fail host.errors.full_messages.join('; ')
+          end
+        end
+      end
+
+      def register_host(host, consumer_params, content_view_environment, activation_keys = [])
+        unless host.new_record?
+          host.save!
+          unregister_host(host, :unregistering => true)
+          host.reload
+        end
+
+        unless activation_keys.empty?
+          content_view_environment ||= lookup_content_view_environment(activation_keys)
+          set_host_collections(host, activation_keys)
+        end
+        fail _('Content View and Environment not set for registration.') if content_view_environment.nil?
+
+        host.save!
+        host.content_facet = plan_content_facet(host, content_view_environment)
+        host.subscription_facet = plan_subscription_facet(host, activation_keys, consumer_params)
+        host.save!
+
+        cp_create = ::Katello::Resources::Candlepin::Consumer.create(content_view_environment.cp_id, consumer_params, activation_keys.map(&:cp_name))
+
+        User.as_anonymous_admin do
+          host = ::Host.find(host.id)
+          unless consumer_params[:facts].blank?
+            ::Katello::Host::SubscriptionFacet.update_facts(host, consumer_params[:facts])
+          end
+        end
+
+        ::Katello.pulp_server.extensions.consumer.create(cp_create[:uuid], display_name: host.name)
+
+        begin
+          set_content_and_subscription_uuids(host, cp_create[:uuid])
+        rescue => e
+          ::Katello::Resources::Candlepin::Consumer.destroy(cp_create[:uuid])
+          raise e
```

Comment:
we should probably delete the pulp consumer as well.

Reply:
good catch, will add